### PR TITLE
issue #1400: don't add duplicate reporter

### LINKED ISSUE #1400: TestNG, Multiple duplicate listener warnings on implementing multiple listener interfaces
Also refer - [Link to Stackoverflow Comments](http://stackoverflow.com/questions/43037690/testng-multiple-duplicate-listener-warnings-on-implementing-multiple-listener-i)

### Expected behavior
**If** i have a single Listener class that implements multiple listeners eg: IAnnotationTransformer, IInvokedMethodListener, ITestListener and IReporter etc...
**Then** the warning indicating multiple listeners must not be displayed if this the act of implementing multiple TestNG listener interface is of no consequence.

### Actual behavior
My Test run successfully BUT i get a warning "[TestNG] [WARN] Ignoring duplicate listener : com.NGtest.RetryListener". This warning is shown for every method call i.e. Listener gets called multiple times per test method invocation, and hence the IGNORED duplicate warning.

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [x] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
**Listener Class:**
```
import java.util.List;
import org.testng.IInvokedMethod;
import org.testng.IInvokedMethodListener;
import org.testng.IReporter;
import org.testng.ISuite;
import org.testng.ITestContext;
import org.testng.ITestListener;
import org.testng.ITestResult;
import org.testng.xml.XmlSuite;

public class RetryListener implements IAnnotationTransformer, IInvokedMethodListener, ITestListener, IReporter {
	@Override
	public void transform(ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod) { }

	@Override
	public void beforeInvocation(IInvokedMethod arg0, ITestResult arg1) { }

	@Override
	public void afterInvocation(IInvokedMethod arg0, ITestResult iTestResult) { }
	
	@Override
	public void onFinish(ITestContext context) { }

	@Override
	public void onStart(ITestContext arg0) { }

	@Override
	public void onTestFailedButWithinSuccessPercentage(ITestResult arg0) {}

	@Override
	public void onTestFailure(ITestResult arg0) { }

	@Override
	public void onTestSkipped(ITestResult arg0) { }

	@Override
	public void onTestStart(ITestResult arg0) { }

	@Override
	public void onTestSuccess(ITestResult arg0) { }

	@Override
	public void generateReport(List<XmlSuite> arg0, List<ISuite> arg1, String arg2) { }
}

```
**Test Class:**
```
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

public class RetryProvider {
    @Test (dataProvider = "datasource")
    public void Test1(int code, String type, ITestContext context){
        System.out.println("Test1(): " + code + " : " + type);
    }

    @DataProvider(name = "datasource")
    public Object[][] getData(){
        Object[][] data = new Object[3][2];
        data[0][0] = 1;
        data[0][1] = "apple";
        data[1][0] = 2;
        data[1][1] = "mango";
        data[2][0] = 1;
        data[2][1] = "guava";
        return data;
    }       
}
```
**Test Xml:**
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="NGBug Suite">
	<listeners><listener class-name="com.NGtest.RetryListener" /></listeners>
	<test name="NGBug Test">
		<classes><class name="com.NGtest.RetryProvider" /></classes>
	</test> <!-- Test -->
</suite> <!-- Suite -->

```
**Test Result:**
```
[TestNG] [WARN] Ignoring duplicate listener : com.NGtest.RetryListener
[TestNG] [WARN] Ignoring duplicate listener : com.NGtest.RetryListener
[TestNG] [WARN] Ignoring duplicate listener : com.NGtest.RetryListener
Test1(): 1 : apple
Test1(): 2 : mango
Test1(): 3 : guava
```

### PR DESCRIPTION
**Pull request by missedone**
Fixes #1400 .

### Did you remember to?

- [ ] Add test case(s)
- [x] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/TestNG.java [hunk 1]
```diff
@@ -1394,7 +1394,9 @@ private SuiteRunner createSuiteRunner(XmlSuite xmlSuite) {
     }
 
     for (IReporter r : result.getReporters()) {
-      addListener(r);
+      if (!m_reporters.containsKey(r.getClass())) {
```
Comment (missedone) [inline src/main/java/org/testng/TestNG.java]: in #createSuiteRunner(), add reporter only if it's not yet existing

Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: Did you consider to replace the line by `maybeAddListener(m_reporters, r.getClass(), r);`?

Comment (missedone) [inline src/main/java/org/testng/TestNG.java]: this is what's i'm trying to avoid, since calling `maybeAddListener` will pring the warning message.

Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: Yes, but we can add a new boolean param in order to determine the behavior.

In fact, the question is more: do we need to add all listener implementations? If yes, we should not check only with `m_reporters`. If no, we should add the listener directly in `m_reporters`.
WDYT about it?

Comment (missedone) [inline src/main/java/org/testng/TestNG.java]: it's for reporter only, PR updated. thx

**Review by juherr — APPROVED** [review]
LGTM 👍 

Ping @cbeust

**Pull request merged**